### DISCUSSION TITLE
Raspberry PI 5 support

### DISCUSSION THREAD

**Discussion by Lenael** [siderolabs/talos#7821] [General]
Few days ago a new Raspberry PI 5 was introduced. It's build based on new Broadcom chip: BCM2712 (full specification: https://www.raspberrypi.com/products/raspberry-pi-5/#specification). It looks really interesting considering its price (80$ for 8GB version). Is there any chance Talos would support this new SBC?

**Comment by smira** [comment]
We are excited about RPi5!

Same answer as for any other SBC: we need upstream Linux support (6.1) and u-boot support. If that lands, we can start accepting patch for RPi5.

**Comment by RobReus** [comment]
@smira is there any update on this? It seems RPi 5 is not yet supported and issue #7978 does not see much activity.

**Reply by frezbo** [reply]
https://github.com/siderolabs/talos/issues/8065 will address all this

**Comment by iamasmith** [comment]
I had a week off and wanted to look at this, there's still a little work to do before sending in a PR (somehow) but I've written drivers for the 2 reset controllers used for brcm,brcmstb-reset and brcm,bcm7216-pcie-sata-rescal (which seem solid) alongside, additionally, a pretty experimental split off driver for brcm,bcm2712-pcie which is the main client driver.
At about 17:00 hours today I was finally able to enum devices on the PCIE lanes off the board.. thankfully the NVME isn't hung off the RP1 southbridge which will need more work but I might give it a go to unlock Ethernet in u-boot/PXE boot etc.
Anyhow here's the device list currently..
```
U-Boot 2024.07-gd655a075 (Aug 30 2024 - 16:56:49 +0100)

DRAM:  1020 MiB (effective 8 GiB)
RPI 5 Model B (0xd04170)
Core:  24 devices, 12 uclasses, devicetree: board
MMC:   mmc@fff000: 0
Loading Environment from FAT... Unable to read "uboot.env" from mmc0:1...
In:    serial,usbkbd
Out:   serial,vidconsole
Err:   serial,vidconsole
mbox: Header response code invalid
bcm2835: Could not query MAC address
Net:   No ethernet found.

reset-controller@119500 SATA/PCIe rescal success
PCIe BRCM2712: link up, 5.0 Gbps x1 (!SSC)
reset-controller@119500 SATA/PCIe rescal success
PCIe BRCM2712: link up, 5.0 Gbps x4 (!SSC)
starting USB...
No USB controllers found
Hit any key to stop autoboot:  0
U-Boot> pci
BusDevFun  VendorId   DeviceId   Device Class       Sub-Class
_____________________________________________________________
00.00.00   0x14e4     0x2712     Bridge device           0x04
01.00.00   0x2646     0x501c     Does not fit any class  0xff
02.01.05   0xffff     0xffff     Does not fit any class  0xff
U-Boot>
```
As you can see the devices show but at this stage aren't reporting a subclass but at this stage after quite a few hours I chose to retire to the 'purple zone' and will look at it again and try to get it nailed before I go back to work on Monday (doing less interesting Cloud Stuff :p).
BTW, yes I based my fork off 2024.07 which is the latest in the Talos beta but I'll ensure the diffs merge and work on current of course.
(btw: I did do some other stuff on my week off :D)
The next step is to work out why the NVME storage is currently showing a sub-class of 0xff (unknown) vs 0x01 (storage) and I think getting that netted will allow the nvme driver to hook up and all may be well, so maybe close to the gate. I suspect that as both devices show this though (including RP1 on bus 02) that it's a matter of enabling some additional clocking. There is some wierdness around a clock called 'sw_pcie' which isn't enabled in the device tree and I had it in code, took it out, added it, took it out.. it may be important but the associated vector looks fishy - but there's code in u-boot to generate safe tries on these things for invalid gets for various things like clocks and resets. It wouldn't surprise me though that the devices on the bus not being clocked could report their IDs but not their class... hmm.

**Comment by iamasmith** [comment]
Thoughts on why this wasn't working were incorrect, I had actually updated the wrong constant based on device specific information in the Linux driver and it proved hard to track down until I re-read the code several times.

I now have correct device information for regular devices attached to the bus, you can ignore 02.1b.05, I know that's the RP1 controller and it needs a u-class driver if/when I/somebody want to get it to work but as you can see the nvme drive is popping up but cannot yet be scanned in a stable fashion.
```
U-Boot> pci
BusDevFun  VendorId   DeviceId   Device Class       Sub-Class
_____________________________________________________________
00.00.00   0x14e4     0x2712     Bridge device           0x04
01.00.00   0x2646     0x501c     Mass storage controller 0x08
02.1b.05   0xffff     0xffff     Does not fit any class  0xff
U-Boot> pci header 01.00.00
  vendor ID =                   0x2646
  device ID =                   0x501c
  command register ID =         0x0006
  status register =             0x0010
  revision ID =                 0x01
  class code =                  0x01 (Mass storage controller)
  sub class code =              0x08
  programming interface =       0x02
  cache line =                  0x08
  latency time =                0x00
  header type =                 0x00
  BIST =                        0x00
  base address 0 =              0x00100004
  base address 1 =              0x00000000
  base address 2 =              0x00000000
  base address 3 =              0x00000000
  base address 4 =              0x00000000
  base address 5 =              0x00000000
  cardBus CIS pointer =         0x00000000
  sub system vendor ID =        0x2646
  sub system ID =               0x501c
  expansion ROM base address =  0x00000000
  interrupt line =              0x00
  interrupt pin =               0x01
  min Grant =                   0x00
  max Latency =                 0x00
U-Boot> nvme scan
"Synchronous Abort" handler, esr 0x96000005, far 0x1b0010001c
elr: 000000000003ada0 lr : 000000000003ad90 (reloc)
elr: 000000003fb7fda0 lr : 000000003fb7fd90
x0 : 0000001b00100000 x1 : 0000000000100000
x2 : 0000001000118000 x3 : 0000000000003e0b
x4 : 0000000000001e01 x5 : 0000000000000001
x6 : 000000003fb7cd98 x7 : 0000000000000028
x8 : 000000003fb7cb60 x9 : 0000000000000000
x10: 0000000000000006 x11: 000000000001869f
x12: 000000003f72dfe8 x13: 000000003f72e140
x14: 0000000000000000 x15: 000000003f74af22
x16: 000000003fb7cb44 x17: 0000000000000000
x18: 000000003f740d60 x19: 000000003f74aec0
x20: 000000003f74a870 x21: 000000003f74a870
x22: 000000003fbc4fc8 x23: 0000000000000002
x24: 000000003fbeb68c x25: 0000000000000000
x26: 000000003f74add0 x27: 0000000000000000
x28: 0000000000000000 x29: 000000003f72dd00

Code: aa0003f3 f8050415 a9000000 f9402660 (b9401c00) 
Resetting CPU ...

resetting ...
```

**Reply by joyshmitz** [reply]
Hi @iamasmith!
Perhaps this repo [Raspberry Pi PCI Express device compatibility database](https://github.com/geerlingguy/raspberry-pi-pcie-devices) will help you, unfortunately I'm not familiar with such subtleties.

The guys seem to know the ins and outs of raspberries.

**Comment by iamasmith** [comment]
@joyshmitz thanks for the post, I'm afraid it's a bit lower level at this point and really about getting the basics working for the bus. As this is the bootloader I don't think there is much value beyond nvme and possibly at some point the peripherals that hang off the RP1 south bridge which is attached to the PCIE bus as well.
I have been looking at old datasheets for other Broadcom devices as much of what is on the PI5 is based on other Broadcom products but it is challenging and there seem to be quite a few unexplained pieces to me at this point and this board is a bit of a challenge in that the only UART available on basic initialisation is the debug port meaning that you have to get the GPIO fired up to have both a serial console off that and an active debug connection using the debug probe which means getting PCIE up so that you can get RP1 going and then enable the UARTs off that - so it's either UART or Debug Port off that one little connector, not both at the same time sadly.

**Comment by prudnitskiy** [comment]
Hi,

are there any updates on this topic? I am eager to run Talos on RPi5 but not sure what is the status of the RPi support in Talos. The issue is resolved, does it mean I'm safe to proceed with the current release?

**Reply by rothgar** [reply]
You can see this issue for more up to date information. There are some workarounds to get a raspberry pi to boot talos but it's not fully supported until the upstream kernel and uboot add support. https://github.com/siderolabs/talos/issues/7978

**Comment by safts** [comment]
If [this](https://www.phoronix.com/news/Raspberry-Pi-RP1-Linux-6.17) is true, does it mean that all we have to do to get Raspberry Pi 5 support is wait for Talos to switch to the 6.17 Linux Kernel?

**Reply by rothgar** [reply]
Talos only uses LTS kernels and 6.17 is supposed to be an LTS kernel. So if it lands in 6.17 then the Talos kernel should support it. We still may be waiting for u-boot support for the rpi5 which currently doesn't fully support the hardware (eg nvme booting).

**Reply by iamasmith** [reply]
and apologies, I had my world a bit realigned last November and this got zero focus. I may try to contribute the reset controllers to u-boot that I did that look pretty much OK but aren't the complete picture and the bcm module definitely needs more work to get the NVME controller functioning properly. Looking at the RPI Kernel patches though at the time.. they were functional but not exactly pretty and in some cases pretty misleading with structs that had values that were not used and actually incorrect for some hardware. This may have changed a lot but in that state I couldn't see a working submission to upstream to support RPI5 happening. TLDR; it all looked pretty hacky at that particular time but times change and they may have been working up to something.

**Reply by bartholdbos** [reply]
Linux 6.17 has been released now and the talos master branch has been updated to 6.17 as well. Does anybody know if u-boot has been fixed to work with the pi 5? Or should talos from master already been working?

**Reply by mglazenborg** [reply]
I just installed Talos v1.12.0-alpha.2, which per the release notes uses Linux 6.17.5. Unfortunatly, it doesn't boot.

However, it seems that SUSE yesterday delivered u-boot for the RPI5.
https://www.phoronix.com/news/SUSE-Raspberry-Pi-5-U-Boot

**Reply by eseiker** [reply]
[u-boot 2026.01 merges linux 6.17 dts](https://github.com/u-boot/u-boot/tree/v2026.01-rc2/dts/upstream/src/arm64/broadcom) so it would have better support for rpi 5. https://github.com/siderolabs/sbc-raspberrypi updated uboot dependency to 2026.01-rc2, so next 1.12.0-beta.0 image **might** work for rpi 5. 🤞

**Reply by 95gabor** [reply]
Talos `v1.12.0-beta.0` didn't boot for me.

**Reply by eseiker** [reply]
my report: https://github.com/siderolabs/sbc-raspberrypi/issues/23#issuecomment-3527204823

you'd need to disable uart to boot, and wait for linux 6.18 lands to get ethernet working...

**Reply by battledash** [reply]
https://github.com/siderolabs/sbc-raspberrypi/pull/64 has been merged, and [v1.12.0-rc.0](https://github.com/siderolabs/talos/releases/tag/v1.12.0-rc.0) is on 6.18. Anyone want to test if it works?

**Reply by WyriHaximus** [reply]
> [siderolabs/sbc-raspberrypi#64](https://github.com/siderolabs/sbc-raspberrypi/pull/64) has been merged, and [v1.12.0-rc.0](https://github.com/siderolabs/talos/releases/tag/v1.12.0-rc.0) is on 6.18. 

Sweet! Didn't get around to asking if 1.12 would be on 6.18 since it's out for a couple of weeks now. But this is very promising 🎉 

> Anyone want to test if it works?

Will have a look tomorrow if someone doesn't meat me to it. Got a fresh 16GB Pi 5 in with NVME SSD, will report back 👍 .

**Reply by 3H-cphillips** [reply]
I've had a 5 just sitting here waiting to test it. I'll see if I can get it added to the cluster sometime in the next 24. Thanks for the work here!

**Reply by virot** [reply]
Not having luck, but I might have played to much with this RPI :). What I do know image factory uses siderolabs/sbc-raspberrypi:v0.1.6, the fix for enabled_uart was merged after, so we still needed to use configTxtAppend.
`configTxtAppend: 'enable_uart=0'`

**Reply by Reversing0148** [reply]
I tested [v1.12.0-rc.0](https://github.com/siderolabs/talos/releases/tag/v1.12.0-rc.0) on 2 separate Compute Module 5's, with both stock and `configTxtAppend: 'enable_uart=0'` unfortunately with no luck. I have no video out on my carrier board, the only way I could check is if the cm5 got an ip assigned via dhcp. It did not, and links on my switch are also down with no packets sent or received.

**Reply by WyriHaximus** [reply]
Tried it using: https://factory.talos.dev/?arch=arm64&board=rpi_generic&bootloader=auto&cmdline-set=true&extensions=-&overlay-options=configTxtAppend%3A+%27enable_uart%3D0%27&platform=metal&target=sbc&version=1.12.0-rc.0

Same result as with the latest `1.11`:
* Pi turns on
* Bootscreenshots some info on what it finds to boot from
* Tries to boot Talos
* Screen goes black
* SSD LED blinks a few times within a second 
* Screen turns off due to no signal
* SSD LED doesn't blink anymore
* Kept it on for several minutes, but nothing

Tried it in the M.2 HAT+, USB NVME adaptor in both USB2 and USB3 ports. All show the same behaviour. Flashed Pi OS on the SSD to make sure everything works, that just boots fine, screen turns on and shows the desktop etc. Did a firmware update, and then repeated the 3 ways of connecting the SSD. Same result.

**Reply by Reversing0148** [reply]
No luck with v1.12.0-rc1 either.

**Reply by patriziobruno** [reply]
I was successful installing talos 1.11.5 on three rpi cm5 and run a cluster on a deskpi super6c following the instructions on https://rcwz.pl/2025-10-04-installing-talos-on-raspberry-pi-5/ The code the blog is based on can be found at https://github.com/talos-rpi5

Having little time to move things farther and wanting to use the Gateway API, I decided to wait for Talos to upgrade  to k8s 1.35. Unfortunately, the authors of talos-rpi5 have been inactive for the past 7 months, hence no Talos-1.12/k8s-1.35.

If anyone can dedicate some time to fork and resume that work, I'll be happy to help as I can.

**Reply by joryirving** [reply]
There is this: https://github.com/talos-on-pi5.
Relevant discussion: https://github.com/siderolabs/sbc-raspberrypi/issues/23

**Reply by battledash** [reply]
I've had success with https://github.com/talos-rpi5 as well, but not when installing over PXE. I compiled iPXE with a script that booted their installer, then their fork of u-boot with a script that loaded iPXE. That worked, but on installation Talos tries to write a boot entry to efivars for systemd-boot, and something in the uboot->ipxe->talos chain was making efivars read-only. The entry isn't really necessary afaik so I removed writing it from their fork of Talos, which got installation working, but it was somehow only reading the installed machineconfig on boot _if a display was connected_. When headless it would boot from disk but be in maintenance mode. So I gave up and decided to wait for official support :stuck_out_tongue:.

**Reply by patriziobruno** [reply]
> There is this: https://github.com/talos-on-pi5.

Thanks! I'll have to figure out how to use this. I guess I should follow official Talos documentation and replace some of the urls with talos-on-pi5's.


> Relevant discussion: [siderolabs/sbc-raspberrypi#23](https://github.com/siderolabs/sbc-raspberrypi/issues/23)

From the discussion, it appears the official Talos is being ported to PI5. I'll wait a couple of weeks and see. There's a good chance the official image will work soon: https://github.com/siderolabs/sbc-raspberrypi/pull/33

**Reply by kees-closed** [reply]
> > There is this: https://github.com/talos-on-pi5.
> 
> Thanks! I'll have to figure out how to use this. I guess I should follow official Talos documentation and replace some of the urls with talos-on-pi5's.
> 
> > Relevant discussion: [siderolabs/sbc-raspberrypi#23](https://github.com/siderolabs/sbc-raspberrypi/issues/23)
> 
> From the discussion it appears the official Talos is being ported to PI5. I'll wait a couple of weeks and see. There's a good chance the official image will soon work soon: [siderolabs/sbc-raspberrypi#33](https://github.com/siderolabs/sbc-raspberrypi/pull/33)

How solid is this assumption? I just bought 2 more Rpi4b's to have all my nodes with the same amount of RAM. But if Rpi5 is really around the corner, then I can still return these 2 and replace them with 16GB Rpi5 nodes. Are all the dependencies for Rpi5 support now available? But just not integrated into Talos? Because if it's in general not available, then you might be selling false hope.

**Reply by patriziobruno** [reply]
@kees-closed you can read siderolabs/sbc-raspberrypi#23 and assess by yourself :)

**Reply by kees-closed** [reply]
> @kees-closed you can read [siderolabs/sbc-raspberrypi#23](https://github.com/siderolabs/sbc-raspberrypi/issues/23) and assess by yourself :)

I've been following that for over a year. But I don't know where you get "soon" from. Sure, maybe somewhere in 2026 support may come. But then the Raspberry Pi 6 is also soon available, maybe with less hurdles. I think I'll wait for the 6 and then upgrade. The support for the Rpi4 is really good and it has an official PoE HAT.

**Reply by mglazenborg** [reply]
Looks like the RPi5 will officially be supported starting 1.12.3 per #12670. The overlay image is already available so might be able to build it yourself with Talos 1.12.2.

**Reply by WyriHaximus** [reply]
@mglazenborg That is awesome! But does it come with nvme support?

**Reply by smira** [reply]
It will not _boot_ from NVMe, but NVMe as a disk for EPHEMERAL and/or user disk should work, but you need to try it out yourself and report back!

**Reply by mglazenborg** [reply]
@smira Ah damn that sucks, what is holding back booting from NVME?

**Reply by smira** [reply]
> @smira Ah damn that sucks, what is holding back booting from NVME?

I guess you need to put some effort into U-boot support for it (that's my understanding). Everything is community based, so if you don't contribute, nothing would really happen

**Reply by WyriHaximus** [reply]
@smira Booting from NVMe is a must for me to be honest.
@mglazenborg For reference: https://github.com/siderolabs/sbc-raspberrypi/issues/23#issuecomment-3789510689

Personally, I lack the skills for U-Boot changes right now, and IIRC, people have been looking into that.

**Reply by smira** [reply]
> @smira Booting from NVMe is a must for me to be honest.

If it's a must for you, I guess you need to do something to make it work - sponsor U-Boot developer or something like that. I don't see the expectation that an open-source project should do the work you need if you're not contributing in any way.

**Reply by WyriHaximus** [reply]
> > @smira Booting from NVMe is a must for me to be honest.
> 
> If it's a must for you, I guess you need to do something to make it work - sponsor U-Boot developer or something like that.

Fully agree, if I can't figure out how to help with the technical side, I'm more than happy to toss some money at it.

> I don't see the expectation that an open-source project should do the work you need if you're not contributing in any way.

I don't have that expectation from U-Boot. Currently, I lack the skills and knowledge to meaningfully contribute besides money and help with testing. Which means I would have to build those skills, which takes time.

To be absolutely clear: I'm not demanding any work from any opensource project, I'm willing to spend time learning and contributing in various ways.

**Reply by WyriHaximus** [reply]
Reached out to one of the U-Boot devs/contributors about this offering to help or sponsor (part of) the work. Will keep you all posted on how that evolves.

**Reply by ivanaguilario** [reply]
@WyriHaximus have there been any developments regarding this?

**Reply by WyriHaximus** [reply]
@ivanaguilario Never received a response to that email. But I signed up for the U-Boot mailing list to keep an eye on it daily, and there might have been a patch merge request to add some or all of the functionality required to boot from NVMe but I haven't confirmed yet if that is the case, and if Talos has updated to that version of U-Boot yet. Will do that probably next week and keep you all posted here.

**Reply by denniseffing** [reply]
@WyriHaximus Do you have any news on the U-Boot update of Talos? 😇

**Reply by WyriHaximus** [reply]
@denniseffing Sorry for the late response, yes and no. Nothing out of the box yet. Currently running a 3rd party overlag from someone on the edge that put something working together with a different kernel. The U-Boot part has been solved, but as it turns out that is only one part of what needs solving. This comment, the while PR is one to keep your eye on, summaries the current situation the best: https://github.com/siderolabs/sbc-raspberrypi/pull/88#issuecomment-5011642792
